### DISCUSSION TITLE
Getting Started

### DISCUSSION THREAD

**Discussion by marcelamtrucco** [Homebrew/discussions#5358] [Getting started]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.21
ORIGIN: https://github.com/Homebrew/brew
HEAD: 82c2e743a5bcea725f9ca1429e3e21c3088ff904
Last commit: 6 days ago
Core tap JSON: 11 May 23:33 UTC
Core cask tap JSON: 11 May 23:33 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Settings.

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 15.1.


Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  picard

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1
  python-charset-normalizer
  python-idna
  python-requests
  python-urllib3

Warning: You have the following deprecated, official taps tapped:
  Homebrew/homebrew-cask-versions
Untap them with `brew untap`.

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python-idna
  python-urllib3
  certifi
  python-charset-normalizer
```

### Description of issue

I have installed Python 3.12.3 using Homebrew in MacOs. I am trying now to install the module requests, using brew install python-requests, and I got the message: Warning: python-requests has been deprecated because it does not meet homebrew/core's requirements for Python library formulae!

==> Downloading https://ghcr.io/v2/homebrew/core/python-requests/manifests/2.31.0
Already downloaded: /Users/marcelamartinez/Library/Caches/Homebrew/downloads/6cb16fa127bf8c35117bb20414003d8bdbd6c27bc2fffc417ad4285eac476f69--python-requests-2.31.0.bottle_manifest.json
==> Fetching python-requests
==> Downloading https://ghcr.io/v2/homebrew/core/python-requests/blobs/sha256:ef53ca288c1f15629dd76ec733aea0111d
Already downloaded: /Users/marcelamartinez/Library/Caches/Homebrew/downloads/cc55a77bc3c7a50965cebcf3134ce59e46f92f0b8a7f9e399dd0f21efbe6b84b--python-requests--2.31.0.arm64_ventura.bottle.tar.gz
==> Reinstalling python-requests 
==> Pouring python-requests--2.31.0.arm64_ventura.bottle.tar.gz
==> Caveats
Additional details on upcoming formula removal are available at:
* https://github.com/Homebrew/homebrew-core/issues/157500
* https://docs.brew.sh/Python-for-Formula-Authors#libraries
* https://docs.brew.sh/Homebrew-and-Python#pep-668-python312-and-virtual-environments
==> Summary
🍺  /opt/homebrew/Cellar/python-requests/2.31.0: 58 files, 457.3KB
==> Running `brew cleanup python-requests`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).

I am confused if the module is not available from Homebrew? when I run import requests in Phyton env, I got the error that the module is not found.

Any help would be appreciated.
Thanks

**Comment by SMillerDev** [comment]
You should probably install it with pip, not with Homebrew since it's going away soon

**Reply by chucker** [reply]
But, FWIW, pip says to use Homebrew…

```
❯ pip3 uninstall requests
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try brew install
    xyz, where xyz is the package you are trying to
    install.
```

(What I ended up doing was `pip3 install --break-system-packages requests`, but that's discouraged…)

**Reply by SMillerDev** [reply]
Pip says you can try to use Homebrew. Sometimes it'll work, sometimes it won't. Using virtual environments will always work.

**Comment by ChristopherSchultz** [comment]
```
    python3 -m venv path/to/venv
    source path/to/venv/bin/activate
    python3 -m pip install xyz
```

Following those instructions:

```
$ python3 -m pip install requests
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try brew install
    xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a Python library that isn't in Homebrew,
    use a virtual environment:
    
    python3 -m venv path/to/venv
    source path/to/venv/bin/activate
    python3 -m pip install xyz
    
    If you wish to install a Python application that isn't in Homebrew,
    it may be easiest to use 'pipx install xyz', which will manage a
    virtual environment for you. You can install pipx with
    
    brew install pipx
    
    You may restore the old behavior of pip by passing
    the '--break-system-packages' flag to pip, or by adding
    'break-system-packages = true' to your pip.conf file. The latter
    will permanently disable this error.
    
    If you disable this error, we STRONGLY recommend that you additionally
    pass the '--user' flag to pip, or set 'user = true' in your pip.conf
    file. Failure to do this can result in a broken Homebrew installation.
    
    Read more about this behavior here: <https://peps.python.org/pep-0668/>

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

Suggestions?

I'm beginning to think that using Python for anything serious isn't tenable in the long-term.

**Reply by ChristopherSchultz** [reply]
Oh, and adding `--break-system-packages` to any command line I tried results in an "unrecognized argument" error.

**Comment by in-ch** [comment]
It can be solved as follows.

```
mkdir ~/.venv 
```

```
python3 -m venv ~/.venv 
```

```
source ~/.venv/bin/activate 
```

```
python3 -m pip install requests
```

One inconvenience is that you'll need to reactivate the virtual environment each time you restart the terminal:

```
source ~/.venv/bin/activate 
```
